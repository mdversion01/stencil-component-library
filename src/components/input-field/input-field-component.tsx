// src/components/input-field/input-field-component.tsx
import { Component, h, Prop, Element, State, Watch } from '@stencil/core';

@Component({
  tag: 'input-field-component',
  styleUrls: ['../layout-styles.scss', '../form-styles.scss', './input-field-styles.scss'],
  shadow: false,
})
export class InputFieldComponent {
  @Element() host!: HTMLElement;

  // ----- Props -----
  @Prop() disabled: boolean = false;
  @Prop({ mutable: true }) formId: string = '';
  @Prop({ mutable: true }) formLayout: '' | 'horizontal' | 'inline' = '';
  @Prop() inputId: string = '';
  @Prop() size: '' | 'sm' | 'lg' = '';
  @Prop() label: string = '';
  @Prop() labelSize: '' | 'sm' | 'lg' = '';
  @Prop() labelHidden: boolean = false;
  @Prop() required: boolean = false;
  @Prop() type: string = 'text';
  @Prop() validation: boolean = false;
  @Prop() validationMessage: string = '';
  @Prop() value: string = '';
  @Prop() placeholder?: string;

  /** Bootstrap grid columns for label when formLayout="horizontal" (default 2) */
  @Prop() labelCol: number = 2;
  /** Bootstrap grid columns for input when formLayout="horizontal" (default 10) */
  @Prop() inputCol: number = 10;

  // Internal
  @State() _resolvedFormId: string = '';
  private inputEl?: HTMLInputElement;

  connectedCallback() {
    const formComponent = this.host.closest('form-component') as any;
    const fcFormId = formComponent?.formId;
    const fcLayout = formComponent?.formLayout;

    if (!this.formId && typeof fcFormId === 'string') this.formId = fcFormId;

    if (!this.formLayout && typeof fcLayout === 'string') {
      const allowed = ['', 'horizontal', 'inline'] as const;
      if ((allowed as readonly string[]).includes(fcLayout)) {
        this.formLayout = fcLayout as '' | 'horizontal' | 'inline';
      }
    }
  }

  componentDidLoad() {
    this.applyFormAttribute();
  }

  @Watch('formId')
  onFormIdChange(newVal: string) {
    this._resolvedFormId = newVal || '';
    this.applyFormAttribute();
  }

  private applyFormAttribute() {
    if (!this.inputEl) return;
    if (this._resolvedFormId) this.inputEl.setAttribute('form', this._resolvedFormId);
    else this.inputEl.removeAttribute('form');
  }

  private camelCase(str: string) {
    if (!str) return '';
    return str
      .replace(/(?:^\w|[A-Z]|\b\w)/g, (word, index) => (index === 0 ? word.toLowerCase() : word.toUpperCase()))
      .replace(/\s+/g, '');
  }

  private handleInput = (ev: Event) => {
    const target = ev.target as HTMLInputElement;
    if (this._resolvedFormId) {
      const form = (target as any).form || document.getElementById(this._resolvedFormId);
      if (form) (target as any).form = form;
    } else {
      (target as any).form = null;
    }
    this.value = target.value;
  };

  /** Compute safe label/input col classes without mutating props. */
  private getComputedCols() {
    // Defaults
    const DEFAULT_LABEL = 2;
    const DEFAULT_INPUT = 10;

    // Parse numbers + clamp to [1..11] for label and input
    const lbl = Number(this.labelCol);
    const inp = Number(this.inputCol);
    const label = Number.isFinite(lbl) ? Math.max(1, Math.min(11, lbl)) : DEFAULT_LABEL;
    const input = Number.isFinite(inp) ? Math.max(1, Math.min(11, inp)) : DEFAULT_INPUT;

    if (this.formLayout === 'horizontal') {
      if (label + input !== 12) {
        // ‚ùó Do NOT mutate props; just log and fall back to defaults
        console.error(
          '[input-field-component] For formLayout="horizontal", label-col + input-col must equal 12 exactly. ' +
            `Received: ${this.labelCol} + ${this.inputCol} = ${Number(this.labelCol) + Number(this.inputCol)}. ` +
            'Falling back to 2/10.'
        );
        return { label: DEFAULT_LABEL, input: DEFAULT_INPUT };
      }
    }

    return { label, input };
  }

  private renderInputLabel(ids: string, labelColClass?: string) {
    const classes = [
      'form-control-label',
      this.required ? 'required' : '',
      this.labelSize === 'sm' ? 'label-sm' : this.labelSize === 'lg' ? 'label-lg' : '',
      this.labelHidden ? 'sr-only' : '',
      this.formLayout === 'horizontal' ? `${labelColClass} no-padding col-form-label` : '',
      this.validation ? 'invalid' : '',
    ]
      .filter(Boolean)
      .join(' ');

    const text = this.formLayout === 'horizontal' || this.formLayout === 'inline' ? `${this.label}:` : this.label;

    return (
      <label class={classes} htmlFor={ids || undefined}>
        {text}
      </label>
    );
  }

  private renderInput(ids: string, names: string) {
    const sizeClass = this.size === 'sm' ? 'basic-input-sm' : this.size === 'lg' ? 'basic-input-lg' : '';
    const classes = ['form-control', this.validation ? 'is-invalid' : '', sizeClass].filter(Boolean).join(' ');

    const placeholder = this.labelHidden
      ? this.label || this.placeholder || 'Placeholder Text'
      : this.label || this.placeholder || 'Placeholder Text';

    return (
      <div>
        <input
          ref={el => (this.inputEl = el as HTMLInputElement)}
          id={ids || undefined}
          name={names || undefined}
          type={this.type || 'text'}
          class={classes}
          placeholder={placeholder}
          value={this.value || undefined}
          aria-label={this.labelHidden ? names : undefined}
          aria-labelledby={names || undefined}
          aria-describedby={this.validation ? 'validationMessage' : undefined}
          disabled={this.disabled}
          required={this.required}
          onInput={this.handleInput}
        />
        {this.validation && this.validationMessage ? (
          <div id="validationMessage" class="invalid-feedback form-text">
            {this.validationMessage}
          </div>
        ) : null}
      </div>
    );
  }

  render() {
    const ids = this.camelCase(this.inputId).replace(/ /g, '');
    const names = this.camelCase(this.label).replace(/ /g, '');

    const outerClass = this.formLayout ? ` ${this.formLayout}` : '';

    const groupClasses = ['form-group'];
    if (this.formLayout === 'horizontal') groupClasses.push('row');
    else if (this.formLayout === 'inline') groupClasses.push('row', 'inline');

    const { label, input } = this.getComputedCols();
    const labelColClass = `col-${label}`;
    const inputColClass = `col-${input}`;

    return (
      <div class={outerClass}>
        <div class={groupClasses.join(' ')}>
          {this.renderInputLabel(ids, labelColClass)}
          {this.formLayout === 'horizontal' ? (
            <div class={inputColClass}>{this.renderInput(ids, names)}</div>
          ) : (
            this.renderInput(ids, names)
          )}
        </div>
      </div>
    );
  }
}
