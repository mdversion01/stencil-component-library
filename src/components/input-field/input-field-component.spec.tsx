// src/components/input-field/input-field-component.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { InputFieldComponent } from './input-field-component';

describe('<input-field-component>', () => {
  it('renders with basic props and a11y attributes', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component label="First Name" input-id="first"></input-field-component>`,
    });

    const root = page.root!;
    const label = root.querySelector('label') as HTMLLabelElement;
    const input = root.querySelector('input') as HTMLInputElement;

    expect(label).toBeTruthy();
    expect(input).toBeTruthy();

    // Stacked layout: label text (no colon)
    expect(label.textContent).toBe('First Name');

    // Linkage: assert the input id (stable in Stencil mock);
    // the label's "for" attribute is inconsistently exposed in the mock DOM.
    expect(input.id).toBe('first');

    // Placeholder prefers label
    expect(input.placeholder).toBe('First Name');

    // a11y basics (label visible => no aria-label)
    expect(input.getAttribute('aria-label')).toBeNull();
    // aria-labelledby uses camelCased label
    expect(input.getAttribute('aria-labelledby')).toBe('firstName');
    expect(input.getAttribute('aria-describedby')).toBeNull();
  });

  it('binds value and updates on input (silences Prop immutability warn)', async () => {
    const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});
    try {
      const page = await newSpecPage({
        components: [InputFieldComponent],
        html: `<input-field-component label="Email" input-id="email" value="test@site.com"></input-field-component>`,
      });

      const comp = page.rootInstance as InputFieldComponent;
      const input = page.root!.querySelector('input') as HTMLInputElement;

      // Initial binding
      expect(input.value).toBe('test@site.com');

      // Simulate typing
      input.value = 'new@site.com';
      input.dispatchEvent(new Event('input', { bubbles: true }));
      await page.waitForChanges();

      // Component mutates its own prop (component code does this); assert observed state
      expect(comp.value).toBe('new@site.com');
      expect(input.value).toBe('new@site.com');
    } finally {
      warnSpy.mockRestore();
    }
  });

  it('applies and removes the form attribute when formId changes', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component label="Zip" input-id="zip"></input-field-component>`,
    });

    // Add a real form so the attribute targets something
    const form = document.createElement('form');
    form.id = 'myForm';
    document.body.appendChild(form);

    const comp = page.rootInstance as InputFieldComponent;
    const input = page.root!.querySelector('input') as HTMLInputElement;

    // Initially no form attr
    expect(input.getAttribute('form')).toBeNull();

    // Set formId -> watcher applies attribute
    comp.formId = 'myForm';
    await page.waitForChanges();
    expect(input.getAttribute('form')).toBe('myForm');

    // Clear formId -> attribute removed
    comp.formId = '';
    await page.waitForChanges();
    expect(input.getAttribute('form')).toBeNull();

    document.body.removeChild(form);
  });

  it('renders validation message and aria-describedby when invalid', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component label="Username" input-id="user" validation validation-message="This field is required."></input-field-component>`,
    });

    const input = page.root!.querySelector('input') as HTMLInputElement;
    const msg = page.root!.querySelector('#validationMessage') as HTMLDivElement;

    expect(input.classList.contains('is-invalid')).toBe(true);
    expect(input.getAttribute('aria-describedby')).toBe('validationMessage');
    expect(msg.textContent).toContain('This field is required.');
  });

  it('respects disabled and required flags', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component label="Code" input-id="code" disabled required></input-field-component>`,
    });

    const input = page.root!.querySelector('input') as HTMLInputElement;
    expect(input).toHaveAttribute('disabled');
    expect(input).toHaveAttribute('required');
  });

  it('uses compact responsive col specs in horizontal layout', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component
               label="City"
               input-id="city"
               form-layout="horizontal"
               label-cols="xs-12 sm-4"
               input-cols="xs-12 sm-8"
             ></input-field-component>`,
    });

    const row = page.root!.querySelector('.form-group.row') as HTMLElement;
    expect(row).toBeTruthy();

    const labelEl = row.querySelector('label') as HTMLLabelElement;
    expect(labelEl).toBeTruthy();
    expect(labelEl.className).toMatch(/col-12/);
    expect(labelEl.className).toMatch(/col-sm-4/);

    const inputWrapper = labelEl.nextElementSibling as HTMLElement;
    expect(inputWrapper).toBeTruthy();
    expect(inputWrapper.className).toMatch(/col-12/);
    expect(inputWrapper.className).toMatch(/col-sm-8/);

    // In horizontal/inline layouts, the label ends with a colon
    expect(labelEl.textContent).toBe('City:');
  });

  it('adds aria-label when label is hidden (labelHidden=true)', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component label="Phone" label-hidden input-id="phone"></input-field-component>`,
    });

    const input = page.root!.querySelector('input') as HTMLInputElement;
    // When labelHidden, aria-label uses camelCased label
    expect(input.getAttribute('aria-label')).toBe('phone');
    expect(input.placeholder).toBe('Phone');
  });

  // ---------------- SNAPSHOTS ----------------

  it('matches snapshot (default stacked layout)', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component
               label="First Name"
               input-id="first"
               value="Ada"
             ></input-field-component>`,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });

  it('matches snapshot (horizontal with validation)', async () => {
    const page = await newSpecPage({
      components: [InputFieldComponent],
      html: `<input-field-component
               label="Email"
               input-id="email"
               form-layout="horizontal"
               label-cols="xs-12 sm-3"
               input-cols="xs-12 sm-9"
               validation
               validation-message="Please enter a valid email."
               value=""
             ></input-field-component>`,
    });
    await page.waitForChanges();
    expect(page.root).toMatchSnapshot();
  });
});
