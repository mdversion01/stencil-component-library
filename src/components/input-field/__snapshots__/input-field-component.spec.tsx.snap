// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<input-field-component> matches snapshot (default stacked layout) 1`] = `
<input-field-component input-id="first" label="First Name" value="Ada">
  <div>
    <div class="form-group">
      <label class="form-control-label" htmlfor="first">
        First Name
      </label>
      <div>
        <input aria-labelledby="firstName" class="form-control" id="first" name="firstName" placeholder="First Name" type="text" value="Ada">
      </div>
    </div>
  </div>
</input-field-component>
`;

exports[`<input-field-component> matches snapshot (horizontal with validation) 1`] = `
<input-field-component form-layout="horizontal" input-cols="xs-12 sm-9" input-id="email" label="Email" label-cols="xs-12 sm-3" validation="" validation-message="Please enter a valid email." value="">
  <div class="horizontal">
    <div class="form-group row">
      <label class="col-12 col-form-label col-sm-3 form-control-label invalid no-padding" htmlfor="email">
        Email:
      </label>
      <div class="col-12 col-sm-9">
        <div>
          <input aria-describedby="validationMessage" aria-labelledby="email" class="form-control is-invalid" id="email" name="email" placeholder="Email" type="text">
          <div class="form-text invalid-feedback" id="validationMessage">
            Please enter a valid email.
          </div>
        </div>
      </div>
    </div>
  </div>
</input-field-component>
`;
